pkg/providers/gateway/translation/gateway_tcproute_test.go

```diff
@@ -0,0 +1,196 @@
+	svcInformer := informersFactory.Core().V1().Services().Informer()
+	svcLister := informersFactory.Core().V1().Services().Lister()
+	epLister, epInformer := kube.NewEndpointListerAndInformer(informersFactory, false)
+	apisixClient := fakeapisix.NewSimpleClientset()
+	apisixInformersFactory := apisixinformers.NewSharedInformerFactory(apisixClient, 0)
+
+	_, err := client.CoreV1().Endpoints("test").Create(context.Background(), endpoints, metav1.CreateOptions{})
+	assert.Nil(t, err)
+	_, err = client.CoreV1().Services("test").Create(context.Background(), svc, metav1.CreateOptions{})
+	assert.Nil(t, err)
+
+	tr := &translator{
+		&TranslatorOptions{
+			KubeTranslator: translation.NewTranslator(&translation.TranslatorOptions{
+				EndpointLister: epLister,
+				ServiceLister:  svcLister,
+				ApisixUpstreamLister: kube.NewApisixUpstreamLister(
+					apisixInformersFactory.Apisix().V2beta3().ApisixUpstreams().Lister(),
+					apisixInformersFactory.Apisix().V2().ApisixUpstreams().Lister(),
+				),
+				APIVersion: config.DefaultAPIVersion,
+			}),
+		},
+	}
+
+	processCh := make(chan struct{})
+	svcInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			processCh <- struct{}{}
+		},
+	})
+	epInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			processCh <- struct{}{}
+		},
+	})
+
+	stopCh := make(chan struct{}, 2)
+	defer close(stopCh)
+	go svcInformer.Run(stopCh)
+	go epInformer.Run(stopCh)
+	cache.WaitForCacheSync(stopCh, svcInformer.HasSynced)
+	cache.WaitForCacheSync(stopCh, epInformer.HasSynced)
+
+	return tr, processCh
+}
+
+func TestTranslateGatewayTCPRoute(t *testing.T) {
+	refKind := func(str gatewayv1alpha2.Kind) *gatewayv1alpha2.Kind {
+		return &str
+	}
+	refNamespace := func(str gatewayv1alpha2.Namespace) *gatewayv1alpha2.Namespace {
+		return &str
+	}
+	refPortNumber := func(i gatewayv1alpha2.PortNumber) *gatewayv1alpha2.PortNumber {
+		return &i
+	}
+	refInt32 := func(i int32) *int32 {
+		return &i
+	}
+	tr, processCh := mockTCPRouteTranslator(t)
+	<-processCh
+	<-processCh
+
+	tcpRoute := &gatewayv1alpha2.TCPRoute{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "tcp_route",
+			Namespace: "test",
+		},
+		Spec: gatewayv1alpha2.TCPRouteSpec{
+			Rules: []gatewayv1alpha2.TCPRouteRule{
+				{
+					BackendRefs: []gatewayv1alpha2.BackendRef{
+						{
+							BackendObjectReference: gatewayv1alpha2.BackendObjectReference{
+								Kind:      refKind("Service"),
+								Name:      "svc",
+								Namespace: refNamespace("test"),
+								Port:      refPortNumber(80),
+							},
+							Weight: refInt32(100),
+						},
+					},
+				},
+			},
+		},
+	}
+
+	tctx, err := tr.TranslateGatewayTCPRouteV1Alpha2(tcpRoute)
+	assert.Nil(t, err)
+
+	assert.Equal(t, 1, len(tctx.StreamRoutes))
+	assert.Equal(t, 1, len(tctx.Upstreams))
+
+	r := tctx.StreamRoutes[0]
+	u := tctx.Upstreams[0]
+
+	// Metadata
+	// FIXME
```

Comment:
Fix what?

Reply:
Er, it was copied from httproute. I was negligent.